### DISCUSSION TITLE
Errors when building firmware: "Error: unknown shorthand flag: 'n' in -n"

### DISCUSSION THREAD

**Discussion by eurorackworld** [shanteacontrols/OpenDeck#194] [Q&A] [answered]
I've been trying to build the Opendeck firmware myself and didn't succes after many tries.

1) I tried on my Mac and everything went fine up until the 'make'' command. That command threw many errors 
2) I tried Ubuntu via WIndows / WSL. again, everything went fine up until the 'make' command. That command threw many errors
(I used Docker, but probably due to my missing knowledge that also didn't work out)

I tried different config files , but most of them ended up on the same error. Especially on Ubuntu this error always was shown:
![image](https://user-images.githubusercontent.com/29734704/221377743-6ef145f1-3f93-4062-85a6-a87049275c27.png)

As I have seen the same error 90% of the times on my Mac as well, it looks like I'm doing something wrong. Maybe forgetting something critical when using Linux. I just follow the guide mentioned here https://github.com/shanteacontrols/OpenDeck/wiki/Building-the-OpenDeck-firmware and that all goes fine, up until the final 'make' of the firmware.

Paradajz already helped me a lot by building a few initial builds and giving me suggestions. Maybe this problem above is recognized, or even someone else has it as well, so posting it publicly to see if someone can help but also to get this issue online in case anybody else has it.

It feels like I just miss something, a setting, a command, no idea what, google didn't help me either..

Oh and yes I'm in the :~/OpenDeck/src$ directory when using the command

**Comment by paradajz** [comment]
Which Ubuntu version?

**Reply by eurorackworld** [reply]
![image](https://user-images.githubusercontent.com/29734704/221378433-c1afef5c-7d90-4266-af82-0c3fe5406df6.png)

**Comment by eurorackworld** [comment]
As soon as I have run the command one time, for example for rooibos as target, the next time when I use the make command with the same target it shows this (and it keeps doing it if I again use the same command)

![image](https://user-images.githubusercontent.com/29734704/221378609-cac3f467-7488-4ea9-bc94-66e5b88d3ba3.png)

**Comment by paradajz** [comment, marked as answer]
Can you post the output of:

`dasel --version`

**Reply by eurorackworld** [reply]
Just one line

![image](https://user-images.githubusercontent.com/29734704/221378855-e60a41e0-9615-4b55-9923-197f1fbf0a98.png)

**Reply by paradajz** [reply]
Ah. Run this on your Ubuntu machine:

```bash
cd
wget https://github.com/TomWright/dasel/releases/download/v1.27.3/dasel_linux_amd64
chmod +x dasel_linux_amd64
sudo mv dasel_linux_amd64 /usr/local/bin/dasel
```

Run `dasel --version` afterwards. It should read `dasel version v1.27.3`.

Then in `OpenDeck/src` directory, run `make clean` followed by `make`. Post the output

**Reply by eurorackworld** [reply]
First I got

![image](https://user-images.githubusercontent.com/29734704/221379039-b4e91d20-bf80-4b1f-9d2a-614812b6988b.png)

That missing unzip I see for the first time, so installed unzip.

then again make clean and make

![image](https://user-images.githubusercontent.com/29734704/221379149-dd8e8056-025d-4d33-8b7d-687957204ec4.png)
(etc , a lot of building stuff..)

now running make TARGET=rooibos again and it's doing a lot!

I'll another make TARGET=pico as well, and let you know output of both

**Reply by eurorackworld** [reply]
so first 'make TARGET=rooibos' 

![image](https://user-images.githubusercontent.com/29734704/221379261-5e8b1cea-3654-4aa2-92ce-8364d8b552ee.png)
followed by a lot of the same 'image' notices, continued by building a lot of stuff and ended with
![image](https://user-images.githubusercontent.com/29734704/221379284-cfcce43e-e840-4083-94cd-7b5adcd99639.png)

then 'make TARGET=pico', no image notices, but a lot of building ending with
![image](https://user-images.githubusercontent.com/29734704/221379324-e99a283d-b8d8-4a32-90c3-49c1c6e5f26b.png)

these both look fine right? And I do have a UF2 file for the pico it seems :)!

Looks like or the unzip addition fixed it, or the dasel commands did it (think both were needed)

**Reply by paradajz** [reply]
> so first 'make TARGET=rooibos'
> (...) followed by a lot of the same 'image' notices (...)

You can ignore those. Rooibos target has touchscreen config file, the build system detects that and tries to chop up images into icons. The images in this repository are stored using [Git LFS](https://git-lfs.com/) so they are not downloaded by default since nobody other than me needs that. It's not crucial for the build.

> then 'make TARGET=pico', no image notices, but a lot of building ending with ![image](https://user-images.githubusercontent.com/29734704/221379324-e99a283d-b8d8-4a32-90c3-49c1c6e5f26b.png)

Looks good, yes. Lots of build output is expected - the OpenDeck firmware is composed of bootloader, application, an PC app which generates default configuration for all parameters and another PC app which generates .sysex firmware update file.

> Looks like or the unzip addition fixed it

`unzip` was needed for the default target which is currently `af_ib_nrf52840` (Adafruit ItsyBitsy nRF52840 Express). It's the default since it's the first target in `config/target` directory. nRF targets need nRF SDK which is downloaded as .zip archive. After that the build system extracts it with `unzip` tool which wasn't present on your system, hence the complaining (`unzip` tool was missing in the dependency list for Ubuntu on Wiki page).

> or the dasel commands did it (think both were needed)

`dasel` is the tool which parses MCU and target config files. The command in the instructions downloaded the latest version which uses different flags, so you were getting the errors. I've updated the documentation to use fixed dasel version (the one I've specified above).

**Reply by eurorackworld** [reply]
Awesome, looks like that works. Going to try reinstalling and checking stuff on my mac as well as I did have a bit of the same errors.

But it looks like your learned the man how to fish!! (instead of giving him a fish:P). 

Major thanks!

**Reply by paradajz** [reply]
> Going to try reinstalling and checking stuff on my mac as well as I did have a bit of the same errors.

Just uninstall current `dasel` version first since you've installed it with `brew`:

`brew uninstall dasel`

**Reply by eurorackworld** [reply]
> Ah. Run this on your Ubuntu machine:
> 
> ```shell
> cd
> wget https://github.com/TomWright/dasel/releases/download/v1.27.3/dasel_linux_amd64
> chmod +x dasel_linux_amd64
> sudo mv dasel_linux_amd64 /usr/local/bin/dasel
> ```
> 
> Run `dasel --version` afterwards. It should read `dasel version v1.27.3`.
> 
> Then in `OpenDeck/src` directory, run `make clean` followed by `make`. Post the output

So on my Windows setup (using Bootcamp on my Mac), everything works fine, Awesome!

Maybe we could solve the Mac issue as well? If not, no problem, but maybe it is as simple as on Windows..

`dasel --version`  --> dasel version 2.1.1

'make clean' (or any other make command) now says 'Toolchain.mk:39: *** Required package not found: objcopy.  Stop.'

Weird, as this was different before, so I did do something on my Mac what changed it. 

Trying to do the procedure from step 1 again...
